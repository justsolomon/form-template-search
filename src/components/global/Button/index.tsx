import Spinner from '../Spinner';
import styles from './button.module.scss';

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  /** Classname to apply to button to override default styles */
  className?: string;

  /** Boolean showing the loading state of the button */
  isLoading?: boolean;

  /** Boolean showing the disabled state of the button */
  isDisabled?: boolean;

  /** Determines if the button is an icon button */
  iconButton?: boolean;

  /** Variant of the button. Defaults to `solid` */
  variant?: 'solid' | 'ghost' | 'transparent';

  leftIcon?: React.ReactNode;
  rightIcon?: React.ReactNode;
  children: React.ReactNode;
}

function Button({
  isDisabled,
  isLoading,
  children,
  className = '',
  iconButton,
  variant = 'solid',
  leftIcon,
  rightIcon,
  ...rest
}: ButtonProps) {
  return (
    <button
      className={`${styles['button']} ${className} ${
        styles['button--' + variant]
      } ${iconButton ? styles['button--icon'] : ''} ${
        isLoading ? styles['button--loading'] : ''
      }`}
      disabled={isDisabled || isLoading}
      data-testid="button"
      {...rest}
    >
      {leftIcon ? (
        <span className={styles['button__left-icon']}>{leftIcon}</span>
      ) : null}

      <>
        {isLoading ? (
          <Spinner size="24px" color="#ffffff" data-testid="button__spinner" />
        ) : (
          children
        )}
      </>

      {rightIcon ? (
        <span className={styles['button__right-icon']}>{rightIcon}</span>
      ) : null}
    </button>
  );
}

export default Button;
