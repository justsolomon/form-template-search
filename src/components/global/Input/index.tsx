import styles from './input.module.scss';

interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  leftIcon?: string | React.ReactNode;
  rightIcon?: string | React.ReactNode;
  containerClass?: string;
}

const Input = ({
  leftIcon,
  rightIcon,
  containerClass,
  ...rest
}: InputProps) => {
  return (
    <div className={`${styles['input-group']} ${containerClass}`}>
      {leftIcon ? (
        <div className={styles['input-group__left-element']}>{rightIcon}</div>
      ) : null}

      <input
        className={`${styles['input-group__input']} ${
          leftIcon ? styles['input-group__input--has-left'] : ''
        } ${rightIcon ? styles['input-group__input--has-right'] : ''}`}
        {...rest}
      />

      {rightIcon ? (
        <div className={styles['input-group__right-element']}>{rightIcon}</div>
      ) : null}
    </div>
  );
};

export default Input;
