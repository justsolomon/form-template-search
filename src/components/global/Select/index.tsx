import { ReactComponent as DropdownIcon } from 'assets/vectors/arrow-down.svg';
import styles from './select.module.scss';

interface SelectProps extends React.SelectHTMLAttributes<HTMLSelectElement> {
  label: string;
  options: string[];
}

const Select = ({ label, options, ...rest }: SelectProps) => {
  return (
    <div className={styles['select-group']}>
      <label className={styles['select-group__label']}>{label}</label>

      <select
        className={styles['select-group__select']}
        {...rest}
        data-testid="select"
      >
        {options.map((option, index) => {
          return (
            <option value={option} key={index}>
              {option}
            </option>
          );
        })}
      </select>

      <div className={styles['select-group__dropdown-icon']}>
        <DropdownIcon />
      </div>
    </div>
  );
};

export default Select;
