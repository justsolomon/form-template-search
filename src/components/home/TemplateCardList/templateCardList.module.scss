@import 'styles/variables';

.template-card-list {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 65px;

  @media #{$media-desktop} {
    grid-gap: 50px;
  }

  @media #{$media-desktop--sm} {
    grid-gap: 30px;
  }

  @media #{$media-desktop--sm} {
    grid-gap: 30px;
  }

  @media #{$media-tablet} {
    grid-template-columns: repeat(2, 1fr);
  }

  @media #{$media-mobile--sm} {
    grid-template-columns: repeat(1, 1fr);
    grid-gap: 20px;
  }
}
