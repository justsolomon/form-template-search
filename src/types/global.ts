export type TemplateCategory = 'Health' | 'E-commerce' | 'Education';
export type TemplateOrder = 'Default' | 'Ascending' | 'Descending';

export interface Template {
  name: string;
  created: Date;
  category: TemplateCategory[];
  description: string;
  link: string;
}
