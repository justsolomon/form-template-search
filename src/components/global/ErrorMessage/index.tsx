import { ReactComponent as ErrorIcon } from 'assets/vectors/error.svg';
import { ReactComponent as RetryIcon } from 'assets/vectors/retry.svg';
import Button from '../Button';
import styles from './errorMessage.module.scss';

interface IErrorMessage {
  error: string;
  retryRequest: () => void;
}

function ErrorMessage({ retryRequest, error }: IErrorMessage) {
  return (
    <div className={styles['error-message']}>
      <ErrorIcon />
      <p
        className={styles['error-message__description']}
        data-testid="error-message__description"
      >
        {error === 'TypeError: Failed to fetch'
          ? 'Looks like you lost your connection. Please check it and try again'
          : error}
      </p>
      <Button
        leftIcon={<RetryIcon />}
        onClick={retryRequest}
        className={styles['error-message__button']}
      >
        Try again
      </Button>
    </div>
  );
}

export default ErrorMessage;
