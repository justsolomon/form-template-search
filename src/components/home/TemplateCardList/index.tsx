import { Template } from 'types/global';
import TemplateCard from '../TemplateCard';
import styles from './templateCardList.module.scss';

interface TemplateCardListProps {
  templates: Template[];
}

const TemplateCardList = ({ templates }: TemplateCardListProps) => {
  return (
    <ul
      className={styles['template-card-list']}
      data-testid="template-card-list"
    >
      {templates.map((template, index) => (
        <TemplateCard {...template} key={index} />
      ))}
    </ul>
  );
};

export default TemplateCardList;
