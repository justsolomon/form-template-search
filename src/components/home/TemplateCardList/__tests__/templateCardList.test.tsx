import { render, screen } from '@testing-library/react';
import { Template } from 'types/global';
import TemplateCardList from '..';

const templates: Template[] = new Array(5).fill({
  category: ['Health', 'E-commerce', 'Education'],
  created: '2022-02-14T11:28:57.155463',
  description: 'Consectetur ullamco cupidatat veniam, irure',
  link: 'https://formpl.us/templates',
  name: 'Test template',
});

describe('<TemplateCardList />', () => {
  it('should render the TemplateCardList component correctly', () => {
    const { asFragment } = render(<TemplateCardList templates={templates} />);

    expect(asFragment).toMatchSnapshot();
  });
  it('should render all the templates passed into it', () => {
    render(<TemplateCardList templates={templates} />);

    const templateList = screen.queryByTestId('template-card-list');

    expect(templateList?.querySelectorAll('li').length).toBe(5);
  });
});
