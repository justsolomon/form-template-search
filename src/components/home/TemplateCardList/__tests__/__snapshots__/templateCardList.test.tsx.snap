// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateCardList /> should render the TemplateCardList component correctly 1`] = `[Function]`;
